flink-kubernetes/src/main/java/org/apache/flink/kubernetes/highavailability/KubernetesLeaderElectionDriver.java

```diff
@@ -0,0 +1,279 @@
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+
+import static org.apache.flink.kubernetes.utils.Constants.LABEL_CONFIGMAP_TYPE_HIGH_AVAILABILITY;
+import static org.apache.flink.kubernetes.utils.Constants.LEADER_ADDRESS_KEY;
+import static org.apache.flink.kubernetes.utils.Constants.LEADER_SESSION_ID_KEY;
+import static org.apache.flink.kubernetes.utils.KubernetesUtils.checkConfigMaps;
+import static org.apache.flink.kubernetes.utils.KubernetesUtils.getLeaderInformationFromConfigMap;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * {@link LeaderElectionDriver} implementation for Kubernetes. The active leader is elected using Kubernetes.
+ * The current leader's address as well as its leader session ID is published via Kubernetes ConfigMap.
+ * Note that the contending lock and leader storage are using the same ConfigMap. And every component(e.g.
+ * ResourceManager, Dispatcher, RestEndpoint, JobManager for each job) will have a separate ConfigMap.
+ */
+public class KubernetesLeaderElectionDriver implements LeaderElectionDriver {
+
+	private static final Logger LOG = LoggerFactory.getLogger(KubernetesLeaderElectionDriver.class);
+
+	private final Object lock = new Object();
+
+	private final FlinkKubeClient kubeClient;
+
+	private final Executor ioExecutor;
+
+	private final String configMapName;
+
+	private final String lockIdentity;
+
+	private final KubernetesLeaderElector leaderElector;
+
+	// Labels will be used to clean up the ha related ConfigMaps.
+	private final Map<String, String> configMapLabels;
+
+	private final LeaderElectionEventHandler leaderElectionEventHandler;
+
+	private final KubernetesWatch kubernetesWatch;
+
+	private final FatalErrorHandler fatalErrorHandler;
+
+	@GuardedBy("lock")
+	private volatile CompletableFuture<Void> leaderElectorRunFuture;
+
+	@GuardedBy("lock")
+	private volatile boolean running;
+
+	public KubernetesLeaderElectionDriver(
+			FlinkKubeClient kubeClient,
+			Executor ioExecutor,
+			KubernetesLeaderElectionConfiguration leaderConfig,
+			LeaderElectionEventHandler leaderElectionEventHandler,
+			FatalErrorHandler fatalErrorHandler) {
+
+		this.kubeClient = checkNotNull(kubeClient, "Kubernetes client");
+		this.ioExecutor = checkNotNull(ioExecutor, "IO Executor");
+
+		this.leaderElectionEventHandler = checkNotNull(leaderElectionEventHandler, "LeaderElectionEventHandler");
+		this.fatalErrorHandler = checkNotNull(fatalErrorHandler);
+
+		this.configMapName = leaderConfig.getConfigMapName();
+		this.lockIdentity = leaderConfig.getLockIdentity();
+		this.leaderElector = kubeClient.createLeaderElector(leaderConfig, new LeaderCallbackHandlerImpl());
+		this.configMapLabels = KubernetesUtils.getConfigMapLabels(
+			leaderConfig.getClusterId(), LABEL_CONFIGMAP_TYPE_HIGH_AVAILABILITY);
+
+		leaderElectorRunFuture = CompletableFuture.runAsync(leaderElector::run, ioExecutor);
```

Comment:
Both for ZooKeeper and Kubernetes, it could happen. But it seems that the old behavior is also like this and we do not find a critical issue.
Maybe we do not need to check the `running` in the `LeaderElectionDriver` when calling the callbacks. Since when the driver is closed, all the callbacks will be ignored.

Reply:
I will add test for what I have said above. `DefaultLeaderElectionService` should ignore the late event(happen after closed).